pkg/util/sketch/hash.go

```diff
@@ -0,0 +1,52 @@
+package sketch
+
+import "hash/fnv"
+
+// Copyright (c) 2012-2017  Dustin Sallings <dustin@spy.net>
```

Comment:
Out of curiosity why did you copy the function here?

Reply:
It's an optimization for the hasing requirement of count min sketch, which is that for each row of the sketch (`d` depth), you need a separate hash function. These hash functions need to be pairwise independent, meaning that for each row of the sketch, passing the same input to each rows hash function should yield a different row, meaning that for each row we end up in a different column.

It's linked in our design doc and also in a comment in this file, but [this](http://www.eecs.harvard.edu/~michaelm/postscripts/tr-02-05.pdf) paper outlines that you can achieve the same pairwise independence using only two hash functions and for each row using the two hashes for the `event` and an extra little bit of math (hopefully [this](https://github.com/grafana/loki/pull/9882/files#diff-4df9b704f7d9ac192074c6b300f408db456bc8c96bcb4bef15d1bf058bee8af5R34-R37) link works, otherwise look for the `getPos` function).

The tl; dr is that by using only 2 hash functions rather than `d` hash functions, only hashing twice rather than once for each row of the sketch, we do less work when inserting something into a sketch. The hash computation is the majority of CPU time for inserting something into the sketch, and we just cut that in half if not better.

Reply:
Thanks for the lengthy reply. What I meant is why is the hash function not imported from a package since the license at the top suggests is third party code.

Reply:
we only need the one hashing optimization, the rest of our structures are slightly different and we may want to make modifications to them, such as the underlying type for each counter in a count min sketch

for reference, the hash function is from here: https://github.com/dustin/go-probably/blob/master/count.go#L40-L69